core/src/main/java/com/linecorp/armeria/client/logging/ConnectionPoolLoggingListener.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2015 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.logging;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.ConnectionPoolListener;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.TextFormatter;
+import com.linecorp.armeria.common.util.Ticker;
+
+import io.netty.util.AttributeKey;
+import io.netty.util.AttributeMap;
+
+/**
+ * Decorates a {@link ConnectionPoolListener} to log the connection pool events.
+ */
+public class ConnectionPoolLoggingListener implements ConnectionPoolListener {
+    private static final Logger logger = LoggerFactory.getLogger(ConnectionPoolLoggingListener.class);
+
+    private static final AttributeKey<Long> OPEN_NANOS =
+            AttributeKey.valueOf(ConnectionPoolLoggingListener.class, "OPEN_NANOS");
+
+    private final AtomicInteger activeChannels = new AtomicInteger();
+    private final Ticker ticker;
+
+    /**
+     * Creates a new instance with a {@linkplain Ticker#systemTicker() system ticker}.
+     */
+    public ConnectionPoolLoggingListener() {
+        this(Ticker.systemTicker());
+    }
+
+    /**
+     * Creates a new instance with an alternative {@link Ticker}.
+     *
+     * @param ticker an alternative {@link Ticker}
+     */
+    public ConnectionPoolLoggingListener(Ticker ticker) {
+        this.ticker = requireNonNull(ticker, "ticker");
+    }
+
+    @Override
+    public void connectionOpen(SessionProtocol protocol,
+                               InetSocketAddress remoteAddr,
+                               InetSocketAddress localAddr,
+                               AttributeMap attrs) throws Exception {
+        final int activeChannels = this.activeChannels.incrementAndGet();
+        attrs.attr(OPEN_NANOS).set(ticker.read());
```

Comment:
This line can be in the `if` statement as well?

Reply:
Fixed. Also added some defensive code in `connectionClosed` just in case log level changes at runtime.